### PR TITLE
Portal more sandboxed

### PR DESCRIPTION
**Pull request by alexlarsson**
This adds new features to the self-sandbox portal as required by webkit per:
https://github.com/flatpak/flatpak/issues/3241

There is a new sandbox-flag with support or sharing display(wayland+x11), sound and gpu, as well as allowing dbus and a11y access to the sandbox.

There is also a new pair of expose functions that allow you to expose any directory you have access to in your sandbox.

See the various commit comments or details.

### CODE REVIEW

**Comment by alexlarsson** [comment]
flatpak-spawn example code to use this is here: https://github.com/flatpak/flatpak-xdg-utils/pull/24

File: common/flatpak-context.c [hunk 1]
```diff
@@ -1959,8 +1960,13 @@ flatpak_context_add_bus_filters (FlatpakContext *context,
   flatpak_bwrap_add_arg (bwrap, "--filter");
   if (app_id && session_bus)
     {
-      flatpak_bwrap_add_arg_printf (bwrap, "--own=%s.*", app_id);
-      flatpak_bwrap_add_arg_printf (bwrap, "--own=org.mpris.MediaPlayer2.%s.*", app_id);
+      if (sandboxed)
```
Comment (TingPing) [inline common/flatpak-context.c]: Is this logic backwards?

File: portal/flatpak-portal.c [hunk 2]
```diff
@@ -593,7 +748,33 @@ handle_spawn (PortalFlatpak         *object,
...
+      g_ptr_array_add (flatpak_argv, g_strdup ("--sandbox"));
+
+      if (sandbox_flags & FLATPAK_SPAWN_SANDBOX_FLAGS_SHARE_DISPLAY)
+        {
+          if (sockets != NULL && g_strv_contains ((const char * const *) sockets, "wayland"))
+            g_ptr_array_add (flatpak_argv, g_strdup ("--socket=wayland"));
+          if (sockets != NULL && g_strv_contains ((const char * const *) sockets, "fallback-x11"))
+            g_ptr_array_add (flatpak_argv, g_strdup ("--socket=fallback-x11"));
+          if (sockets != NULL && g_strv_contains ((const char * const *) sockets, "x11"))
+            g_ptr_array_add (flatpak_argv, g_strdup ("--socket=x11"));
```
Comment (TingPing) [inline portal/flatpak-portal.c]: Should this also have `ipc`?

**Comment by TingPing** [comment]
Could we also get a `--sandbox --share-network`?

File: app/flatpak-builtins-run.c [hunk 3]
```diff
@@ -66,7 +67,10 @@ static GOptionEntry options[] = {
   { "log-session-bus", 0, 0, G_OPTION_ARG_NONE, &opt_log_session_bus, N_("Log session bus calls"), NULL },
   { "log-system-bus", 0, 0, G_OPTION_ARG_NONE, &opt_log_system_bus, N_("Log system bus calls"), NULL },
   { "log-a11y-bus", 0, 0, G_OPTION_ARG_NONE, &opt_log_a11y_bus, N_("Log accessibility bus calls"), NULL },
-  { "no-a11y-bus", 0, 0, G_OPTION_ARG_NONE, &opt_no_a11y_bus, N_("Don't proxy accessibility bus calls"), NULL },
+  { "no-a11y-bus", 0, G_OPTION_FLAG_REVERSE, G_OPTION_ARG_NONE, &opt_a11y_bus, N_("Don't proxy accessibility bus calls"), NULL },
+  { "a11y-bus", 0, 0, G_OPTION_ARG_NONE, &opt_a11y_bus, N_("Proxy accessibility bus calls (default except when sandboxed)"), NULL },
+  { "no-sesion-bus", 0, G_OPTION_FLAG_REVERSE, G_OPTION_ARG_NONE, &opt_a11y_bus, N_("Don't proxy session bus calls"), NULL },
```
Comment (matthiasclasen) [inline app/flatpak-builtins-run.c]: typo here: no-sesion-bus should be no-session-bus

File: app/flatpak-builtins-run.c [hunk 4]
```diff
@@ -66,7 +67,10 @@ static GOptionEntry options[] = {
...
+  { "session-bus", 0, 0, G_OPTION_ARG_NONE, &opt_session_bus, N_("Proxy session bus calls (default except when sandboxed)"), NULL },
```
Comment (matthiasclasen) [inline app/flatpak-builtins-run.c]: these new options need to be documented somewhere.

File: common/flatpak-context.c (hunk 1 above)
Comment (alexlarsson) [reply common/flatpak-context.c]: Yeah, it look like it :)

File: portal/flatpak-portal.c (hunk 2 above)
Comment (alexlarsson) [reply portal/flatpak-portal.c]: Probably..

**Pull request merged**